### ISSUE TITLE
pressing CTRL-C to interrupt an scp I get ^C%

### ISSUE THREAD

**Issue by cridemichel** [romkatv/powerlevel10k#197]
If I issue the command:
```shell
> scp <a_file> <some_server>:
```
and I interrupt this command by CTRL-L
I get a line with ^C%
how can I get rid of that?

thank you Cristiano

N.B. with robbyrusell default theme this does not happen

**Comment by romkatv** [comment]
Can you reproduce this issue by typing and interrupting the same command from `zsh -df`?

**Comment by cridemichel** [comment]
no in that case I do not get any character

**Comment by romkatv** [comment]
> I forgot a stty -echoctl in the .zshrc

What does that look like?

**Comment by romkatv** [comment]
To clarify, could you post the part of your `~/.zshrc` that calls `stty -echoctl`?

**Comment by cridemichel** [comment]
> To clarify, could you post the part of your `~/.zshrc` that calls `stty -echoctl`?

I have put 
```zsh
# do not echo control characters in hat notation (e.g. ^C)
stty -echoctl
```
to completely suppress echoing of control chars, but using robbyrussell theme
this is not needed, while with powerlevel10k it is if I want to get rid of ^C%

**Comment by romkatv** [comment]
What do you get with `ZSH_THEME=''`? Does it work like `robbyrussell` or like `powerlevel10k`?

**Comment by cridemichel** [comment]
like robbyrussell

**Comment by romkatv** [comment]
Weird. As far as I know, `echoctl` is the default. That's why you get `Ctrl-C` printed if you interrupt `cat` in `zsh -df`, or in `bash` or in anything. Powerlevel10k doesn't issue any terminal commands, so `Ctrl-C` gets printed as usual, unless you explicitly disable it with `stty`.

Could you verify whether `echoctl` is indeed the default in your terminal? You can do this by adding `setopt no_global_rcs no_rcs; return` at the top of `~/.zshenv`, restarting your terminal, typing `cat` and hitting Ctrl-C.

**Comment by cridemichel** [comment]
> setopt no_global_rcs no_rcs; return
```zsh
MacBook-Pro-4% cat
^C
MacBook-Pro-4%
```

**Comment by romkatv** [comment]
It seems like it's the default in your terminal, just like in all of my terminals.

I've no idea what `robbyrussell` does but it doesn't seem like powerlevel10k is doing anything weird. Does this match your understanding of the situation?

**Comment by cridemichel** [comment]
> > setopt no_global_rcs no_rcs; return
> 
> ```shell
> MacBook-Pro-4% cat
> ^C
> MacBook-Pro-4%
> ```

**Comment by cridemichel** [comment]
I think so... but is it not possible to replicate robbyrussell behavior in powerlevel10k?

**Comment by romkatv** [comment]
You mean to call `stty -echoctl` in powerlevel10k? If that's what you want, you can call it yourself. It seems inappropriate to change terminal configuration from a theme.

**Comment by romkatv** [comment]
Does everything the way you want if you add `stty -echoctl` to `~/.zshrc`?

**Comment by cridemichel** [comment]
robby russell theme is the following bunch of lines:
```zsh
local ret_status="%(?:%{$fg_bold[green]%}➜ :%{$fg_bold[red]%}➜ )"
PROMPT='${ret_status} %{$fg[cyan]%}%c%{$reset_color%} $(git_prompt_info)'

ZSH_THEME_GIT_PROMPT_PREFIX="%{$fg_bold[blue]%}git:(%{$fg[red]%}"
ZSH_THEME_GIT_PROMPT_SUFFIX="%{$reset_color%} "
ZSH_THEME_GIT_PROMPT_DIRTY="%{$fg[blue]%}) %{$fg[yellow]%}✗"
ZSH_THEME_GIT_PROMPT_CLEAN="%{$fg[blue]%})"
```
and it is not unsetting the echoctl flag...
I am not sure it is generally safe to unset such a flag

**Comment by romkatv** [comment]
> Does everything work the way you want if you add `stty -echoctl` to `~/.zshrc`?

Could you answer this question?

> and it is not unsetting the echoctl flag...

We cannot see it from the code you posted as it's not self-contained.

The unusual and surprising thing is not that `^C` gets printed when you are using powerlevel10 -- it's the default behavior of your terminal -- but that it *doesn't* get printed in some circumstances. What breaks printing of `^C` on your system I cannot say. You can try to figure it out by bisecting your config. For example, you mentioned that enabling `robbyrussell` theme is enough to suppress the echo of `^C`. You can try to verify this by adding the following snippet to the top of your `~/.zshenv` and restarting your terminal.

```zsh
if [[ -o interactive ]]; then
  ZSH=~/dotfiles/oh-my-zsh
  ZSH_THEME=robbyrussell
  plugins=()
  source $ZSH/oh-my-zsh.sh
fi
setopt no_rcs
return
```

If this doesn't suppress `^C`, the culprit lies elsewhere in your config.

If you can give me a self-contained config that reproduces the suppression of `^C`, I can try to figure out what causes it.

**Comment by cridemichel** [comment]
doing as you suggested does not suppress the ^C% hence culprit is elsewhere...
but my impression is that simply robbyrussell (but also agnoster) theme simply redraws
the promptline on top of ^C%, hth

**Comment by romkatv** [comment]
> doing as you suggested does not suppress the ^C% hence the culprit is elsewhere...

I expected as much.

My guess is that when you are experiencing the suppression of `^C`, it's the result of a bug introduced by something in your config. It's possible that it also triggers on `reset-prompt`. You can verify it by hitting `esc-x` and then typing `reset-prompt`. If this moves your prompt one line up, there is definitely a bug somewhere.

> hth

To be clear, I'm not debugging this. I have no means to do that as this problem cannot be reproduced from the description you provided. I also don't think there is anything wrong with powerlevel10k.

I can help you with debugging this if you can figure out what in your config triggers the suppression of `^C`. That is, if you tell me what I need to do to enable `^C` suppression on my own machine.

**Comment by cridemichel** [comment]
If I create an empty .zshrc I get ^C%, hence it is not related to my specific zsh configuration.
In principle to reproduce the problem you should use an empty .zshrc, but according 
to what you said you can not reproduce the problem this way, right?
if I execute reset-prompt nothing's shown...

**Comment by romkatv** [comment]
> If I create an empty .zshrc I get ^C%

Me too.

> In principle to reproduce the problem you should use an empty .zshrc, but according to what you said you can not reproduce the problem this way, right?

The default behavior of any terminal I know of is to display `^C` when the user interrupts `cat` with `Ctrl-C`. However, when you are using `robbyrussell` theme, this doesn't happen for you. If you can tell me what I need to do to also prevent `^C` from showing up, then I might be able to debug your issue.

We've already established that the following config is not enough, neither for you nor for me, to suppress `^C`. In other words, `^C` does get printed when using this config.

```zsh
ZSH=~/dotfiles/oh-my-zsh
ZSH_THEME=robbyrussell
plugins=()
source $ZSH/oh-my-zsh.sh
```

Which self-contained config disables the printing of `^C`?

**Comment by cridemichel** [comment]
the option which fix the problem with robbyrussell theme is the following one:

```zsh
unsetopt PROMPT_SP
```
if I remove this from my .zshrc, I get ^C% also with robbyrussell  theme

**Comment by romkatv** [comment]
> the option which fix the problem with robbyrussell theme is the following one:

What do you mean by the problem? Could you rephrase this statement by specifying the configuration you use, the actions you perform (such as typing commands) and the console output you observe?

> ```shell
> unsetopt PROMPT_SP
> ```

Do I understand it correctly that you have `unsetopt PROMPT_SP` in your `~/.zshrc`?

Note that unsetting `prompt_sp` is dangerous as it will truncate the output of commands without your knowing. For example, try typing `zsh -df --nopromptsp` (or just use your own config with `rubbyrussell` theme) and then copy-paste this command:

```zsh
echo -n 'Hello\nBye'
```

Or this:

```zsh
echo -n 'Hello\nBye' >/tmp/message && cat /tmp/message
```

**Comment by cridemichel** [comment]
> > the option which fix the problem with robbyrussell theme is the following one:
> 
> What do you mean by the problem? Could you rephrase this statement by specifying the configuration you use, the actions you perform (such as typing commands) and the console output you observe?
> 
> > ```shell
> > unsetopt PROMPT_SP
> > ```
> 
> Do I understand it correctly that you have `unsetopt PROMPT_SP` in your `~/.zshrc`?
> 
> Note that unsetting `prompt_sp` is dangerous as it will truncate the output of commands without your knowing. For example, try typing `zsh -df --nopromptsp` (or just use your own config with `rubbyrussell` theme) and then copy-paste this command:
> 
> ```shell
> echo -n 'Hello\nBye'
> ```
> 
> Or this:
> 
> ```shell
> echo -n 'Hello\nBye' >/tmp/message && cat /tmp/message
> ```

you are right, it is unsafa to use
```zsh
unsetopt PROMPT_SP
```
but I discovered that I can set 
```zsh
PROMPT_EOL_MARK=""
```
to get rid at least of the % chars when I press CTRL-C.
I will stick to this configuration thank you very much for your help!

**Comment by romkatv** [comment]
I've pushed https://github.com/romkatv/powerlevel10k/commit/8ace14beb068b662088f38ccba33b78f8e79bffd that will make powerlevel10k behave the same way as robbyrussell theme with your config. I don't think it's a good idea to unset `prompt_sp` but it shouldn't be the job of powerlevel10k to enforce good practices. Users who want to shoot themselves in the foot should be allowed to. Thanks for reporting this issue!

> ```
> PROMPT_EOL_MARK=""
```

This option isn't as bad as `prompt_sp`. It hides useful information but not too much.

**Comment by cridemichel** [comment]
> I've pushed [8ace14b](https://github.com/romkatv/powerlevel10k/commit/8ace14beb068b662088f38ccba33b78f8e79bffd) that will make powerlevel10k behave the same way as robbyrussell theme with your config. I don't think it's a good idea to unset `prompt_sp` but it shouldn't be the job of powerlevel10k to enforce good practices. Users who want to shoot themselves in the foot should be allowed to. Thanks for reporting this issue!
> 
> > ```
> > PROMPT_EOL_MARK=""
> > ```
> 
> This option isn't as bad as `prompt_sp`. It hides useful information but not too much.

which information can hide this option?

**Comment by romkatv** [comment]
> which information can hide this option?

Type this:

```zsh
echo -n 'Hello' >/tmp/message && cat /tmp/message
```

And then this:

```zsh
echo 'Hello' >/tmp/message && cat /tmp/message
```

(Note that the first `echo` has `-n` while the second doesn't.)

With `PROMPT_EOL_MARK=""` the output is identical even though the files have different content. It's common to accidentally create files without `\n` at the end, and they become difficult to notice and fix with `PROMPT_EOL_MARK`. This may or may not be important to you.

FWIW, I use this in my dotfiles:

```zsh
PROMPT_EOL_MARK='%K{red} %k'
```

**Comment by cridemichel** [comment]
```zsh
PROMPT_EOL_MARK='%K{red} %k'
```
ok probably I will use this, anyway the actual behavior of powerlevel10k is not identical to robbyrussell theme, in that if I add
```zsh
unsetopt prompt_sp prompt_cr 
```
according to the dicussion which I found here   https://unix.stackexchange.com/questions/167582/why-zsh-ends-a-line-with-a-highlighted-percent-symbol, and I issue the command
```zsh
echo -n 'Hello\nBye'
```
with robbyrussell I get:
```zsh
Hello
Bye
```
while with powershell I get only "hello"

**Comment by cridemichel** [comment]
PROMPT_EOL_MARK='%K{red} %k'
ok probably I will use this, anyway the actual behavior of powerlevel10k is not identical to robbyrussell theme, in that if I add

unsetopt prompt_sp prompt_cr 
according to the dicussion which I found here https://unix.stackexchange.com/questions/167582/why-zsh-ends-a-line-with-a-highlighted-percent-symbol <https://unix.stackexchange.com/questions/167582/why-zsh-ends-a-line-with-a-highlighted-percent-symbol>, and I issue the command

echo -n 'Hello\nBye'
with robbyrussell I get:

Hello
Bye
while with powershell I get only "hello"

**Comment by romkatv** [comment]
Unsetting `prompt_cr` is even crazier than unsetting `prompt_sp`. But you are right, if I decided to enable one crazy thing, I should allow all of them. I've submitted https://github.com/romkatv/powerlevel10k/commit/1547336bb6012189632f43ccb6730b02e7425ad0 to respect `no_prompt_cr`.

To see how terrible `no_prompt_cr` is, type `zsh -df --nopromptsp --nopromptcr` and then `echo -n Hello`. Or you can add `unsetopt prompt_sp prompt_cr` to your `~/.zshrc` and use any single-line prompt.

**Comment by cridemichel** [comment]
with robbyrussell theme adding
```zsh 
unsetopt prompt_sp prompt_cr
```
and issuing 
```zsh
echo -n "Hello"
```
I get "Hello", while powershell10k overwrite it

**Comment by cridemichel** [comment]
after upgrading my powerlevel10k repo to latest commit it works as robbyrussell theme,
i.e.  
```zsh
echo -n "hello" 
```
just prints "hello"

**Comment by romkatv** [comment]
> after upgrading my powerlevel10k repo to latest commit,
> by unsetting both prompt_sp and prompt_cr, it works as robbyrussell theme,

I'm glad. Please don't take it as encouragement to unset `prompt_sp` and `prompt_cr`, although you are of course free to do so.

**Comment by cridemichel** [comment]
initially I thought the % as a misbehavior of my zsh but now I see it is a specific feature to show clearly unterminated lines, hence I think I will just use your nicer PROMPT_EOL_MARK without
unsetting prompt_sp prompt_cr. In any case it is good to have powerlevel10k behaving as all other themes

**Comment by romkatv** [comment]
> initially I thought the % as a misbehavior of my zsh but now I see it is a specific feature to show clearly unterminated lines, hence I think I will just use your nicer PROMPT_EOL_MARK without
> unsetting prompt_sp prompt_cr.

Good call. The default behavior of zsh is indeed superb. It's better than in other shells. You can customize `PROMPT_EOL_MARK` to make it look a bit nicer but that's basically all you can do without making things worse.

> In any case it is good to have powerlevel10k behaving as all other themes

Maybe. Or maybe not. If someone unsets `prompt_cr` or `prompt_sp`, it's most likely because they are confused. Powerlevel10k in this case used to set `prompt_cr` and `prompt_sp` back to their sane values, thereby fixing things and making zsh behave well. After my "fixes" those users will have their prompts broken for indefinite amount of time.

It's the same with `auto_name_dirs`. Anyone who sets this option is mistaken. I could unset it for them in powerlevel10k but I don't. Same story with invalid locale, invalid `TERM`, `DISABLE_MAGIC_FUNCTIONS=true` in Oh-My-Zsh (most users do not set this option and suffer needlessly), or `ZSH_AUTOSUGGEST_MANUAL_REBIND=1` (likewise, most users don't know that they should set this option). In all these cases I could easily fixes issues but then someone would complain that powerlevel10k behaves differently from something else, and I'd rather not deal with those complaints.

**Comment by cridemichel** [comment]
for me you can now undo last two commits if you feel more comfortable with that, 
since I am convinced that customizing PROMPT_EOL_MARK  is enough :-)

cheers Cristiano

**Comment by romkatv** [comment]
Naah, I didn't make those commits just to satisfy you. As I mentioned above, I generally don't try to fix unrelated issues for users. If they want to do crazy things, let them. I'm not sure it's the right approach but it's the principle I can follow consistently.

**Comment by cridemichel** [comment]
ok good, btw ZSH_AUTOSUGGEST_MANUAL_REBIND=1 it is another great tip too!!

**Comment by romkatv** [comment]
> ok good, btw ZSH_AUTOSUGGEST_MANUAL_REBIND=1 it is another great tip too!!

You are welcome :grin: 

You already have `DISABLE_MAGIC_FUNCTIONS=true`? It'll make copy-pasting into your prompt 1000 times faster.

**Comment by cridemichel** [comment]
> Il giorno 2 set 2019, alle ore 14:33, Roman Perepelitsa <notifications@github.com> ha scritto:
> 
> ok good, btw ZSH_AUTOSUGGEST_MANUAL_REBIND=1 it is another great tip too!!
> 
> You are welcome 😁
> 
> You already have DISABLE_MAGIC_FUNCTIONS=true? It'll make copy-pasting into your prompt 1000 times faster.
> 
yes sure!
